src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -521,85 +498,42 @@ private bool TryInsert(TKey key, TValue value, InsertionBehavior behavior)
 
             if (comparer == null)
             {
-                if (typeof(TKey).IsValueType)
+                Debug.Assert(typeof(TKey).IsValueType);
```

Comment:
`if (typeof(TKey).IsValueType)` was evaluated by the JIT as constant and the whole block was dead-code eliminated. Change the condition above to `if  (typeof(TKey).IsValueType && comparer == null)` to maintain this optimization? There are likely going to be regressions in Dictionary micro-benchmarks without this optimization.

(multiple places)